[#7158] Determine RTLD flags from DT_FLAGS_1 for loading plugins

SwooshyCueb opened the pull request:
Addresses #7158 for main.
Prerequisite for irods/irods_rule_engine_plugin_python#157

SwooshyCueb commented:
Still running tests, but they're far enough along that I think it's safe to mark this PR as ready for review.

SwooshyCueb commented:
I think we should document this functionality and how to use it, but I'm not sure where

trel commented on lib/core/include/irods/irods_load_plugin.hpp:
```diff
@@ -94,6 +101,274 @@ namespace irods
+            }
+            else {
+                // clang-format off
+                logger::warn({
```
confirming this is `warn` on purpose.

SwooshyCueb replied:
I used `warn` because the situation is less severe than than the cases we use `error` for above, but either one makes sense here to me.

trel replied:
got it - up to you.

trel commented on lib/core/include/irods/irods_load_plugin.hpp:
```diff
@@ -94,6 +101,274 @@ namespace irods
+            // TODO: some platforms do not require a section table for libraries
+            if (e_shoff == 0) {
+                // clang-format off
+                logger::warn({
```
same, warn vs error.

SwooshyCueb replied:
`warn` is intentional here. The absence of a section table doesn't necessarily mean there's something wrong, just that we haven't taught the server how to find `DT_FLAGS_1` without one.

trel replied:
great.

korydraughn commented on lib/core/include/irods/irods_load_plugin.hpp:
```diff
@@ -94,6 +101,274 @@ namespace irods
 
     } // resolve_plugin_path
 
+    /**
+     * \brief attempt to parse DT_FLAGS_1 to identify flags to pass to dlopen()
+     *
+     * \since 4.3.1
+     **/
```
Let's stick with `///` since that appears to have become the convention.

Also, missing `\param[in] _plugin_path description ...`, `\throws`, and `\returns`.

SwooshyCueb replied:
To my understanding, the only potential source of exceptions from this function would be the logging in the exception handlers, so I'm not sure what to put for `\throws`.

korydraughn replied:
The logger uses nlohmann::json, fmtlib, etc.

I think std::exception is good enough. At the very least, anyone calling this needs to catch std::exception to be safe.

If it helps, I've started leaving out **why** a particular exception may be thrown. I find that the message can quickly become incorrect over time and therefore it's safer to just let the dev know which exceptions can be thrown. They can interrogate the `.what()` for the reason.

alanking commented on lib/core/include/irods/irods_load_plugin.hpp:
```diff
@@ -94,6 +101,274 @@ namespace irods
+                // clang-format off
+                logger::error({
+                    {"plugin_path", _plugin_path},
+                    {"message", "Could not parse DT_FLAGS_1: failed to open plugin file"}
```
Our regular/default log messages use the key "log_message". Should we use the same key for consistency?

I realize that by now this may not be consistent everywhere, but figured I'd ask.

SwooshyCueb replied:
Good catch. I used `log_message` in PREP so I'll use that here too.

korydraughn commented on lib/core/include/irods/irods_load_plugin.hpp:
```diff
@@ -94,6 +101,274 @@ namespace irods
+     * \since 4.3.1
+     **/
+    // NOLINTNEXTLINE(readability-function-cognitive-complexity) this is technically a parser
+    inline int get_plugin_rtld_flags(const std::string& _plugin_path) noexcept
```
Consider dropping the use of `noexcept`.

The logger can throw exceptions.

korydraughn commented on lib/core/include/irods/irods_load_plugin.hpp:
```diff
@@ -94,6 +101,274 @@ namespace irods
+                return rtld_lazy_or_now | rtld_local_or_global | rtld_nodelete;
+            }
+
+            // TODO: some platforms do not require a section table for libraries
```
If the section table isn't available, will that affect the PREP in regard to the DT_FLAGS_1 solution you've built?

SwooshyCueb replied:
Yes. Without a section table, we can't currently get to the dynamic entry with `DT_FLAGS_1`, so the plugin would be loaded with the default flags.

I don't see this coming up any time soon. We presently only support x86_64 on Linux, so there will always be a section table. Nonetheless, I'll be opening an issue covering the TODOs once everything else passes review.

korydraughn replied:
Ok. That works for me.

korydraughn commented on lib/core/include/irods/irods_load_plugin.hpp:
```diff
@@ -94,6 +101,274 @@ namespace irods
+            lib.seekg(e_shoff);
+
+            // some data may be contained in the first entry
+            if ((e_shnum == 0) || (e_shstrndx == SHN_XINDEX)) {
```
For non-const variables, let's invert the operands on each side of `==` to guard against future modifications.
```c++
0 == e_shnum || ...
```

SwooshyCueb replied:
You're gonna have to unwrap that one for me. How does switching the operands guard against future modifications?

trel replied:
just protects against fat-fingering the `==` to an assignment...

https://en.wikipedia.org/wiki/Yoda_conditions

SwooshyCueb replied:
That makes sense now that I think about it. I'll make the changes.

korydraughn commented on lib/core/include/irods/irods_load_plugin.hpp:
```diff
@@ -94,6 +101,274 @@ namespace irods
+                        continue;
+                    }
+                    sh_name = s_hdr_in.sh_name;
+                    s_name = std::string_view(&e_strtbl_vec.data()[sh_name]);
```
So `&e_strtbl_vec[sh_name]` doesn't work?

Is the pointer returned null-terminated?

SwooshyCueb replied:
> So `&e_strtbl_vec[sh_name]` doesn't work?

My understanding is that depends on the implementation of `std::vector`, but I could be mistaken.

> Is the pointer returned null-terminated?

Yes.

alanking reviewed (commented):
Seems fine overall in that nothing really jumped out at me. I might need to go read the ELF spec :) Just saw one thing which may or may not matter.

korydraughn commented on lib/core/include/irods/irods_load_plugin.hpp:
```diff
@@ -94,6 +101,274 @@ namespace irods
+                        dt_flags_1 = de_in.d_un.d_val;
+                    }
+
+                    if ((dt_flags_1 & static_cast<std::uint64_t>(DF_1_NOW)) != 0U) {
```
Just curious. Why `!= 0U` instead of `== DF_1_NOW`?

Are you interested in a subset of the bits being set?

SwooshyCueb replied:
Yes, if any other flags are also set, `== DF_1_NOW` would return false, even if `DF_1_NOW` is set.

korydraughn replied:
Unless I'm misunderstanding, the bitwise-and with `DF_1_NOW` will only return the bits matching `DF_1_NOW`.

SwooshyCueb replied:
oh, I misunderstood your question.

clang-tidy suggested `!= 0U`, so that's what I did. But `== DF_1_NOW` seems more correct. I'll make the changes.

korydraughn commented on lib/core/include/irods/irods_load_plugin.hpp:
```diff
@@ -94,6 +101,274 @@ namespace irods
+        }
+        // NOLINTNEXTLINE(hicpp-signed-bitwise)
+        return rtld_lazy_or_now | rtld_local_or_global | rtld_nodelete;
+    }
```
Let's add a trailing comment containing the name of this function here.

korydraughn commented on lib/core/include/irods/irods_load_plugin.hpp:
```diff
@@ -94,6 +101,274 @@ namespace irods
+     * \since 4.3.1
+     **/
+    // NOLINTNEXTLINE(readability-function-cognitive-complexity) this is technically a parser
+    inline int get_plugin_rtld_flags(const std::string& _plugin_path) noexcept
```
Consider adding comments to the most important parts of this function. It's just nice not having to find documentation for all the scary data structures to understand what is happening here.

korydraughn replied:
Yes. Please mark as resolved.

korydraughn reviewed (commented):
Very nice implementation.

Once approved, please capture the TODOs in issues and link to the relevant parts of the implementation.

korydraughn reviewed (commented):
I think we're in a good spot.

Please squash without the pounds and we'll do one last look over it before approving.

SwooshyCueb commented:
Squashed

korydraughn reviewed (approved):
Please add the pound.

SwooshyCueb commented:
Looks like there may be a problem with Centos 7, so I'm going to hold off until I can figure it out.

SwooshyCueb commented:
The problem was a test I added relying on a Python module that didn't exist yet. I've removed the test in question. https://github.com/irods/irods_rule_engine_plugin_python/pull/157#issuecomment-1623880239

\#'d
